Clarify globs

cjllanwarne opened the pull request:
Clarifies the set of files expected to be returned by the `glob()` function, including that it might return 0, 1, or many files in the output `Array[File]`.

Horneth commented on versions/draft-3/SPEC.md:
````diff
@@ -891,14 +891,16 @@ output {
 }
 ```
 
+The array of `File`s returned matches the bash evaluation of the glob from the execution directory the context of the bash version installed in the docker image running the task. In other words, it contains the files which would be matched by running `ls <glob>` from the task's execution directory.
````
Is something off in the first sentence or is it just me ?

cjllanwarne replied:
Hmm, you're right, it is a little bit strange on second reading. I'll try to break it into a couple of sentences instead

orodeh commented:
One of my users asked to sort the glob results. Can we do that?

cjllanwarne commented:
@orodeh I've added the clarification that the files should be returned in the same order that bash would return them if we call `ls <glob>`. This matches the Cromwell implementation.

rasmuse commented:
Nice that you're cleaning up, @cjllanwarne. This comment is just to check if I really understand you right and if you really mean what you wrote :)

## The output of `ls <glob>`

I'm not sure, but I suspect that the intended behavior is *not* to get the same files as returned by `ls <glob>`. Consider the following:

```bash
~$ mkdir tempdir && cd tempdir
~/tempdir$ touch file1
~/tempdir$ mkdir a_directory another_directory
~/tempdir$ touch a_directory/file2 another_directory/file3
~/tempdir$ ls *
file1

a_directory:
file2

another_directory:
file3
```

In other words, `ls <glob>` runs `ls` on all the files and dirs matched by `<glob>`. And I don't even think it's 100% clear what list of files and dirs should be considered the response here. `file1`, `file2`, `file3`? Or `file1` and the two directories? Or all five items?

### Maybe you mean `echo <glob>`?

Closer to what I would expect is `echo <glob>`:

```bash
~/tempdir$ echo *
a_directory another_directory file1
```

### Including directories?

The only potential problem is that only `file` is a file in the everyday sense of the word, but the other two matches are files in a wider sense which includes directories. Which brings me to a related question (which I can't find an answer to in the spec): does/should the WDL type `File` include directories? If then, I suppose `ls <glob>` can just be replaced by `echo <glob>` in @cjllanwarne's clarification?

## Link to bash `glob` docs?

By the way, I'm sure you know, but for reference, here's a link to the relevant bash docs on globs: https://www.gnu.org/software/bash/manual/bashref.html#Filename-Expansion -- would that be relevant to link from the spec or some other documentation?

cjllanwarne commented:
@rasmuse I think I addressed your comments and we discussed it in that Gitter conversation. Could you just add a thumbs-up here that it's looking good now?

geoffjentry commented:
Actually good you avoid using thumbs for this purpose? It makes vote tallying a bit murkier

rasmuse commented:
Yes, thanks, it's good this way. I'm a little confused about voting though - this issue is not marked "Voting Active" yet, but I'm ready to vote yes on it.

geoffjentry commented:
@rasmuse Right, please don't vote until we announce it ready for voting (which will likely be soon on this one)

geetduggal commented on versions/draft-3/SPEC.md:
````diff
@@ -890,15 +892,37 @@ output {
+```
+Then running `echo a*` in the execution directory would expand to `a1.txt`, `ab.txt`, `a_dir` and `az.txt` in that order. Since `glob()` does not include directories we discard `a_dir` and the result of the WDL glob would be `["a1.txt", "ab.txt", "az.txt"]`.
+
+##### Task portability and non-standard BaSH
````
Minor: is the mixed capitalization here for 'bash'  on purpose?

geetduggal commented:
:+1:

orodeh commented:
👍

mlin commented:
:+1:

aprabhak2 commented:
👍

geoffjentry commented:
:+1:

Passes 5-0
